### PR TITLE
Refactor type definitions

### PR DESCRIPTION
**Pull request by 3846masa**
The latest type definitions is invalid to pass `sign` / `verify` / `decode` options.
And, `reply.jwtSign` isn't defined yet.

This PR rewrites type definitions for usability.

### CODE REVIEW

**Comment by mcollina** [comment]
I'll let @Ethan-Arrowood review this.

File: index.d.ts [hunk 1]
```diff
@@ -1,49 +1,68 @@
...
+      sign: jwt.SignOptions;
+      verify: jwt.VerifyOptions;
+    };
+    secret: jwt.Secret;
+
+    sign(payload: JWTTypes.PayloadType | Buffer, options?: jwt.SignOptions): string;
+    sign(payload: JWTTypes.PayloadType | Buffer, callback: JWTTypes.SignCallback): void;
+    sign(payload: JWTTypes.PayloadType | Buffer, options: jwt.SignOptions, callback: JWTTypes.SignCallback): void;
+
+    verify<Decoded extends JWTTypes.PayloadType = any>(token: string, options?: jwt.VerifyOptions): Decoded;
```
Comment (Ethan-Arrowood) [inline index.d.ts]: What is the purpose of defaulting this to `any` if you are constraining it to `JWTTypes.PayloadType`?

File: index.d.ts [hunk 2]
```diff
@@ -1,49 +1,68 @@
...
-    {
-        jwt: Jwt;
-    }
+import * as fastify from 'fastify';
+import * as http from 'http';
+import * as jwt from 'jsonwebtoken';
+
+declare module 'fastify' {
+  namespace JWTTypes {
+    type PayloadType = object | string;
```
Comment (Ethan-Arrowood) [inline index.d.ts]: Update this to the following intersection: `string | object | Buffer | NodeJS.ReadableStream`
Then you can remove the `| Buffer` from the `sign` method parameters

File: index.d.ts [hunk 3]
```diff
@@ -1,49 +1,68 @@
...
+
+declare interface FastifyJWTOptions {
+  secret: jwt.Secret;
+  decode?: jwt.DecodeOptions;
+  sign?: jwt.SignOptions;
+  verify?: jwt.VerifyOptions;
 }
 
-declare const fastifyJWT: fastify.Plugin<Server, IncomingMessage, ServerResponse, { secret: Secret }>;
+declare const fastifyJWT: fastify.Plugin<any, any, any, FastifyJWTOptions>;
```
Comment (Ethan-Arrowood) [inline index.d.ts]: Concerned with the use of `any` here. Why was this changed from the `http` items? I'm unsure how the `fastify.Plugin` type signature works so maybe something needs to be changed there too.

File: index.d.ts (hunk 1 above)
Comment (3846masa) [reply index.d.ts]: `PayloadType` is `object | string`, but `object | string` type requires type assertion for use.
For example,

```ts
{
  const token = app.jwt.sign({ key: 'value' });
  const decoded = app.jwt.verify(token); // <= object | string
  console.log(decoded.key);
}
```

throw Error like below.

```
Property 'key' does not exist on type 'string | object'.
  Property 'key' does not exist on type 'string'.
```

So if default value type is `object | string`, you should set type assertion.
It is hard to use, I think.

For convenience, I set `any` type as default.
When you want to check decoded type strictly, you can use generics.

```ts
interface Data {
  key: string;
}

{
  const orig: Data = { key: 'value' };
  const token = app.jwt.sign(orig);
  const decoded = app.jwt.verify<Data>(token);
  console.log(decoded.key);
}
```

File: index.d.ts (hunk 2 above)
Comment (3846masa) [reply index.d.ts]: `jsonwebtoken` library's `sign()` allow to pass `object` or ` Buffer` or `string` (See [here](https://github.com/auth0/node-jsonwebtoken#jwtsignpayload-secretorprivatekey-options-callback)).

However, `jsonwebtoken` library's `verify()` and `decode()` returns `string` or `object`, because these use `jwt` library's `decode()` (See [here](https://github.com/brianloveswords/node-jws/blob/ab660dbf716bfbff4eda3f9cf6e318cf7e3bfb29/lib/verify-stream.js#L57-L78)).

File: index.d.ts (hunk 3 above)
Comment (3846masa) [reply index.d.ts]: `HttpServer` / `HttpRequest` / `HttpResponse` in `fastify.Plugin<HttpServer, HttpRequest, HttpResponse, T>` maybe `http2` type (See [here](https://github.com/fastify/fastify/blob/b23a2a3923e6b644e910c89d4cabd1536c81922d/fastify.d.ts#L12-L14)).

However, this plugin don't require there type information, so I set `any` type.

File: index.d.ts (hunk 1 above)
Comment (Ethan-Arrowood) [reply index.d.ts]: I have to disagree about this. You use TypeScript to strictly type your JavaScript code. By allowing things to be `any` like this you're opening yourself back up to the exact problems TypeScript is used to mitigate. If a user wishes to inline their `sign` payload then they can extend the JWT type interface and overload the method `payload` parameter. Furthermore, the `jsonwebtoken` types are strict about method payload types (https://github.com/DefinitelyTyped/DefinitelyTyped/blob/9fc7f6fbbdb741ea15c46b0778dd3595d483da26/types/jsonwebtoken/index.d.ts#L133)

File: index.d.ts (hunk 3 above)
Comment (Ethan-Arrowood) [reply index.d.ts]: Leave this as is for now. I'm working on making the base `fastify` types better so we will introduce a better fix here later on

File: index.d.ts (hunk 2 above)
Comment (Ethan-Arrowood) [reply index.d.ts]: Then in this case I'd rather see multiple type definitions such as `SignPayloadType` and `VerifyPayloadType` etc.

File: index.d.ts (hunk 1 above)
Comment (3846masa) [reply index.d.ts]: I'm not good at English, so I cannot completely understand what you want to say.
But I agree that we should avoid using `any` type.

Which should I do?

1. Revert types (Not use Generics)
2. Remain Generics but remove any type (like below)
    ```ts
    verify<Decoded extends JWTTypes.PayloadType = JWTTypes.PayloadType>(token: string, options?: jwt.VerifyOptions): Decoded;
    ```

File: index.d.ts (hunk 2 above)
Comment (3846masa) [reply index.d.ts]: Ok, I'll rewrite.

File: index.d.ts (hunk 3 above)
Comment (3846masa) [reply index.d.ts]: Ok, revert it.

File: index.d.ts (hunk 1 above)
Comment (Ethan-Arrowood) [reply index.d.ts]: You should keep generics without the assignment
```typescript
verify<Decoded extends JWTTypes.PayloadType>(token: string, options?: jwt.VerifyOptions): Decoded;
```

Comment (3846masa) [reply index.d.ts]: Ok, I'll fix it.

**Comment by 3846masa** [comment]
@Ethan-Arrowood I rewrited, please re-review.

**Comment by Ethan-Arrowood** [comment]
@mcollina am I good to squash and merge or would you like one more review on this?

**Comment by mcollina** [comment]
Go for it!

Il giorno mar 9 apr 2019 alle 18:55 Ethan Arrowood <notifications@github.com>
ha scritto:
